### ISSUE TITLE
Clarifying the mask propogation in np.ma.polyfit in documentation

### ISSUE THREAD

**Issue by indiajoe** [numpy/numpy#9193]
I would like to add an example in the documentation of `np.ma.polyfit` to demonstrate that the 2D mask is collapsed on 1D mask before doing the polyfit of a masked array. The current documentation is not very clear on how `np.ma.polyfit` deals with 2D mask while fitting a 2D y array.

For example.
```
A = np.ma.array([[0, 0, 0],
                          [1, 1, 1],
                          [2, 2, 2],
                          [10, 14, 99]])

x = np.arange(A.shape[0])

# only the mA[3,2] entry is masked in mA below
mA = np.ma.masked_greater(A,90)

print(mA)
# polynomial fit of all three columns of A
np.ma.polyfit(x,A,1)
# Outputs: array([[  3.1,   4.3,  29.8],
                 [ -1.4,  -2.2, -19.2]])

# polynomial fit of all three columns of mA
np.ma.polyfit(x,mA,1)
# Outputs: array([[  1.00000000e+00,   1.00000000e+00,   1.00000000e+00],
                       [  4.10073934e-17,   4.10073934e-17,   4.10073934e-17]])

```
Important Note: Masking one element in the last column affected the polynomial fit of all columns.

**Comment by indiajoe** [comment]
More I think about it, I feel `np.ma.polyfit` should not be combining all the masks into a single union mask. I can't think of any application where one would just want to throw away data points from the fitting of independent columns just because some columns have certain rows masked.

I understand the compromise one has to make is in the computation time, since one will have to loop over all columns to get unique masks for each column.

Wouldn't it be better to slice the columns containing similar mask pattern and pass them as a set to `np.polyfit`?
That way, we will have to loop over only the number of times column masks are different.

If others also agree to that approach, I could implement that and submit a pull request.

**Comment by mhvk** [comment]
@indiajoe - I agree that this seems undesired behaviour. Not sure how easy it is to split over mask sets; maybe easiest is to check if the masking is identical, and, if not, to iterate over the axis.

As always, there is also the worry of breaking backwards compatibility. cc @eric-wieser, who may have suggestions.

**Comment by indiajoe** [comment]
@mhvk 
I was thinking we could something like this 
`np.any(np.diff(mA.mask,axis=1),axis=0)` where mA is the masked 2D array to get an array of boolean which has `True` where ever there is a change in the mask between two adjacent columns.
Then, we can loop over each of those continues set of columns and then `np.hstack` the final results.

That way, we wouldn't end up iterating over individual columns just because one column had a different mask.

**Comment by mhvk** [comment]
If one goes that route, I'd actually construct unique sets of masks and iterate over those, filling in a pre-allocated result for each set. But before we go too far along that route, maybe start with the simple version? Also, we do really need to think about backward compatibility...

**Comment by indiajoe** [comment]
@mhvk 
> By constructing a unique set and filling pre-allocated result

Did you have something like the following procedure in mind.

First create an unique set of the column masks and also their indices for inverse mapping of the location of columns.
```
b = np.ascontiguousarray(mA.mask.T).view(np.dtype((np.void, mA.mask.dtype.itemsize * mA.mask.shape[0])))

_,uniqindx,inversemap_indx =  np.unique(b, return_index=True,return_inverse=True)
```
Then, iterate through the unique column masks, and views of the data using the inversemap_indx.

Finally, populate the results matrix using the inversemap_indx.

Regarding backward compatibility, we are not breaking the cases were all columns have same mask. When they had non-homogenous mask, I wonder anybody would have relied on np.ma.polyfit to union combine all the masks. But I might be wrong. 
Should we make this as a non-default with optional argument flag?

**Comment by mhvk** [comment]
@indiajoe - yes, indeed, although I hadn't specifically considered `np.unique` (which is quite suitable in this case for returning the indices, but a bit unhandy in almost certainly requiring a copy of the mask). The only thing would be to generalize the `.T` to a `.swapaxis(axis, 0)` so one can have an arbitrary axis.

As for backwards compatibility and a possible keyword argument, perhaps best would be to ask on the mailing list -- I do agree that the current behaviour is surprising, so more likely that people got *wrong* results out of it than that they rely on it to get *correct* ones.

**Comment by indiajoe** [comment]
@mhvk 
Thanks, I shall use `.swapaxis`
The reason I thought of using `np.unique` is because it gives the inverse mapping indices as well as unique indices in a single step. Both are needed to reconstruct the results array.

I have just now sent an email to the mailing list asking inputs on the backward compatibility.

**Comment by josef-pkt** [comment]
I would stick with the current behavior as default:

- if we want to compare trends across columns, then it might be more useful to use the same observations, if for example there are correlated shocks/noise/outliers.
- the computational complexity increases a lot. In the worst case, you have to loop over all columns and compute the SVD/least squares solution for each column separately
- https://docs.scipy.org/doc/numpy/reference/generated/numpy.polyfit.html has extra output. Many of those will be different for each column or for each missing/mask pattern and need to return masked or vectorized versions. (V is already 3-D, so there is no difference there in the return shape)

The main question is whether you can do much better than a plain loop for the independent mask version.

extra: Is there a check on the minimum number of observations, which would then vary by column, or is it always just a SVD-regularized solution? Needs special case masked handling for "empty" columns.

disclaimer: I'm not a user of np.ma.polyfit.

**Comment by indiajoe** [comment]
@josef-pkt 
Thanks, very important points. 

1. Yes, that is a good use case for present behaviour. But I feel if such uniform masks were needed, the user should have ideally done that explicitly before passing the data to `ma.polyfit`. Feels like `ma.polyfit` is otherwise doing extra things than its name suggests.
But, I agree we should probably keep that default..
2. Yes, currently we were thinking of clubbing all the columns with similar mask together, so we have to loop through only different types of masks. There might be more clever methods.. but I can't think of any now..
3. This is a big problem. I completely overlooked. The `residual` and `V` can be easily inserted back into the full matrix. But the  `rank, singular_values and rcond` are one per each mask set. Not clear to me how to deal with it. Can't think of a way to pass them back to user without breaking the format of returned stuffs to be different from `np.polyfit`. :disappointed: 

extra: Currently there is no separate check on the minimum number of points in `np.ma.polyfit`. It raises the same warnings `np.polyfit` raises when there are not enough points to fit.

**Comment by eric-wieser** [comment]
> there is also the worry of breaking backwards compatibility.

I don't think we should be changing either `polyfit` or `np.ma.polyfit`, since they both use the bad poly1d ordering.

Unfortunately, `np.polynomial.Polynomial.fit(x, A,1)` doesn't work at all, despite being documented as supported

**Comment by josef-pkt** [comment]
> But I feel if such uniform masks were needed, the user should have ideally done that explicitly before passing the data to ma.polyfit

I think it's difficult for users, they would have to combine the masks of up to 3 arrays (x, y, w) and override the original mask which is presumably element wise. 
Row-wise deletion for missing values is a pretty common default for this kind of statistics or estimation. I had to think for a bit to understand that the change wants to use just a separate and independently vectorized version. (I tried to see what np.ma.cov does, but don't really understand the logic.)

I don't see a problem adding an option, but I wouldn't use it as default, given computational cost and backwards compatibility.

**Comment by hdrake** [comment]
I understand the concerns about computational cost and backwards compatibility, but would it be possible to at least add a note about how the mask propagates in the documentation? I spent a full day trying to debug my code only to realize that numpy.ma.polyfit was acting in a very counter-intuitive way.

**Comment by NanoExplorer** [comment]
The documentation needs to be more explicit about the fact that when performing multiple fits all at once, if one row in one column is masked that will cause that row to be masked in all columns. 

I have recently encountered this problem --- I expected the masks to be independent on each column, which caused me to receive unexpected results.

The statement "Any masked values in x is propagated in y, and vice-versa." in the docs might give a clue to this behavior, but that's not how I interpreted it.

**Comment by Jenteb07** [comment]
Hi, 
We just encountered this same problem and it took us a while to figure out why our values made no sense.
Given that this topic has not been updated in 4 years I would like to revive it.

Currently the documentation does not give any warning to the user regarding the usage of masks and is therefor prone to mistakes.

Thanks in advance.
